modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ClassWrapper.java

```
@@ -0,0 +1,113 @@
+import static javax.lang.model.element.Modifier.STATIC;
+import static org.apache.ignite.internal.configuration.processor.ConfigurationProcessorUtils.sameType;
+
+import java.lang.annotation.Annotation;
+import java.util.List;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.TypeMirror;
+import org.apache.ignite.internal.util.Lazy;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A wrapper around a class (represented by a {@link TypeElement}) with some helper methods.
+ */
+public class ClassWrapper {
+    private final ProcessingEnvironment processingEnvironment;
+
+    private final TypeElement clazz;
+
+    private final List<VariableElement> fields;
+
+    private final Lazy<ClassWrapper> superClass = new Lazy<>(this::computeSuperClass);
+
+    /** Constructor. */
+    public ClassWrapper(ProcessingEnvironment processingEnvironment, TypeElement clazz) {
+        this.processingEnvironment = processingEnvironment;
+        this.clazz = clazz;
+        this.fields = fields(clazz);
+    }
+
+    public TypeElement clazz() {
+        return clazz;
+    }
+
+    public <A extends Annotation> A getAnnotation(Class<A> annotationType) {
+        return clazz.getAnnotation(annotationType);
+    }
+
+    /**
+     * Returns the superclass of the class, or {@code null} if the superclass is {@link Object}.
+     */
+    public @Nullable ClassWrapper superClass() {
+        return superClass.get();
+    }
+
+    /**
+     * Returns the superclass of the class, or throws {@link ConfigurationValidationException} if the superclass is {@link Object}.
+     */
+    public ClassWrapper requiredSuperClass() {
+        ClassWrapper superClass = superClass();
+
+        if (superClass == null) {
+            throw new ConfigurationValidationException(this, "Must have a superclass");
```

I think it's better to just throw `NullPointerException` in the same way as in `java.util.Objects#requireNonNull(T)`, since formally there is no validation here and it should be in another place.

---

If you do that, than in all validation methods that require a superclass, you will have to write:
```
requireSuperclass(clazz);
clazz.getSuperClass();
```
Instead of simply writing `getRequiredSuperClass`, which I find more convenient.

---

`ConfigurationValidationException` is special, because it gets printed in a special way. I agree that this exception being here is not obvious, but actually this method is first called during validation, so if we don't have a superclass, we will fail early during the validation phase with a nice exception. This method is also being called later simply because it is convenient to not have to check for nulls. We could have two separate methods, one to use during validation and one to use outside validation, but why bother?

---

Okay, let's leave it.